### PR TITLE
Ensure alert rules can be passed from multiple workload charms

### PR DESCRIPTION
**Pull request by dstathis**
## Issue
When multiple workload charms were related to this charm, it would only forward alert rules for one of them.

## Solution
Concatenate a list of all alert rules.

## Testing Instructions
Hook up two instances of spring-music and make sure alert rules make it in to prometheus for each of them.

## Release Notes
Fixed an issue where alert rules were not forwarded when multiple workload charms were related.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -196,8 +196,11 @@ def _prometheus_configurations(self):
             configured_jobs.append(job)
 
         alerts = list(self._metrics_providers.alerts().values())
+        alert_groups = {"groups": []}  # type: ignore
+        for entry in alerts:
+            alert_groups["groups"] = alert_groups["groups"] + entry["groups"]
 
-        return {"scrape_jobs": configured_jobs, "alert_rules": alerts[0] if alerts else []}
+        return {"scrape_jobs": configured_jobs, "alert_rules": alert_groups if alerts else []}
```
Comment (rbarry82) [inline src/charm.py:203]: This looks subtly wrong, but I'm waiting for it to deploy to test.

`"alert_rules": alert_groups if alert_groups["groups"] else []}` 

Or:

```python
alert_rules = {"groups: []"} if alerts else {}
...

..., "alert_rules": alert_rules or []}
```

?

The variable naming is a little confusing since it's used as a value in a dict with a different name immediately after, but mostly it's confusing because we return a `dict` if `alerts` isn't empty, and a `list` if it is. It should be the same type returned no matter what.

**Review by rbarry82 — COMMENTED** [review]
A+ full-bore integration

File: src/charm.py (hunk 1 above)
Comment (dstathis) [reply src/charm.py:203]: You are right. I think there is an issue here. Let me see what a charm with no alert rules is supposed to return.

File: src/charm.py [hunk 2]
```diff
@@ -196,8 +196,14 @@ def _prometheus_configurations(self):
             configured_jobs.append(job)
 
         alerts = list(self._metrics_providers.alerts().values())
-
-        return {"scrape_jobs": configured_jobs, "alert_rules": alerts[0] if alerts else []}
+        alert_groups = {"groups": []}  # type: ignore
+        for entry in alerts:
+            alert_groups["groups"] = alert_groups["groups"] + entry["groups"]
```
Comment (sed-i) [inline src/charm.py:201]:
```suggestion
            alert_groups["groups"] += entry["groups"]
```

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -1,16 +1,165 @@
...
+from typing import List, Literal
 
-log = logging.getLogger(__name__)
+import aiohttp
+from pytest_operator.plugin import OpsTest
 
 
-async def get_unit_address(ops_test, app_name: str, unit_num: int) -> str:
-    status = await ops_test.model.get_status()  # noqa: F821
+class Prometheus:
```
Comment (sed-i) [inline tests/integration/helpers.py:11]:
```suggestion
class PrometheusClient:
```

File: src/charm.py [hunk 4]
```diff
@@ -196,8 +196,14 @@ def _prometheus_configurations(self):
...
+
+        return {
+            "scrape_jobs": configured_jobs,
+            "alert_rules": alert_groups if alert_groups["groups"] else {},
+        }
```
Comment (sed-i) [inline src/charm.py:206]: I think some unit tests around this are crucial.

File: src/charm.py (hunk 2 above)
Comment (rbarry82) [reply src/charm.py:201]: It will get inlined anyway, so this is preference and readability.

We could also vote for:

```python
alert_groups["groups"].extend(entry["groups"])

# Or...
alert_groups = {
  "groups": [entry["groups"] for entry in alerts]
```

I like that last one, but it doesn't matter that much.

File: tests/integration/helpers.py (hunk 3 above)
Comment (rbarry82) [reply tests/integration/helpers.py:11]: :-1: for a number of reasons:

* It's a helper in an integration test and the name doesn't matter that much as long as we know what it means
* `PrometheusClient | prometheus_client` is the naming for the official "this application provides a metrics endpoint and this `PrometheusClient | prometheus_client` is full of methods for creating graphs/vectors/etc.
* This code is directly from `tests/integration/helpers.py` in `prometheus-k8s`, and we don't need to diverge there

File: src/charm.py (hunk 4 above)
Comment (rbarry82) [reply src/charm.py:206]: :+1: 

There are no unit tests for any alerting rules at all at the moment.

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+async def test_multiple_workloads_alert_rules(ops_test):
+    old_rules = await get_prometheus_rules(ops_test=ops_test, app_name=PROM_NAME, unit_num=0)
+    await ops_test.model.deploy(
+        "ch:spring-music", application_name="spring-music2", channel="edge"
+    )
+    await ops_test.model.add_relation(APP_NAME, "spring-music2")
+    await ops_test.model.wait_for_idle(status="active")
+    new_rules = await get_prometheus_rules(ops_test=ops_test, app_name=PROM_NAME, unit_num=0)
+    # Ensure that adding a second workload creates new alert rules.
```
Comment (mmanciop) [inline tests/integration/test_charm.py:50]: I think this check deserves an in-depth assertion of what new alerts you expect and how they look like. Several bugs, one nastier than the next, would fulfill this check.

**Review by mmanciop — CHANGES_REQUESTED** [review]
Tests need more detailed expectations

File: tests/integration/test_charm.py (hunk 5 above)
Comment (dstathis) [reply tests/integration/test_charm.py:50]: Discussed in person and decided this is okay.

File: src/charm.py (hunk 1 above)
Comment (dstathis) [reply src/charm.py]: Fixed and pushed.

File: tests/integration/helpers.py (hunk 3 above)
Comment (dstathis) [reply tests/integration/helpers.py:11]: You are correct that this is taken directly from the prometheus repo including the class name. We will keep it as is to match that unless there are major objections.

File: src/charm.py (hunk 4 above)
Comment (dstathis) [reply src/charm.py:206]: Unit test added. Let me know if you would like anything else tested.

**Review by rbarry82 — APPROVED** [review]

File: tests/integration/test_charm.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python3
```
Comment (sed-i) [inline tests/integration/test_charm.py:1]: We probably shouldn't be naming files "test_charm" anymore.
How about `test_charm_rules_reach_prometheus`?

File: tests/integration/test_charm.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    await ops_test.model.deploy(charm_under_test, application_name=APP_NAME)
+    # The charm should be in blocked state if not related to anything.
+    await ops_test.model.wait_for_idle(status="blocked")
+
+
+async def test_relate(ops_test):
+    await asyncio.gather(
+        ops_test.model.deploy("ch:prometheus-k8s", application_name=PROM_NAME, channel="beta"),
+        ops_test.model.deploy("ch:spring-music", application_name=SPRING_NAME, channel="edge"),
+    )
```
Comment (sed-i) [inline tests/integration/test_charm.py:29]: Consider moving the `deploy` to the first function.

File: tests/integration/test_charm.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+async def test_alert_rules_exist(ops_test):
+    rules = await get_prometheus_rules(ops_test=ops_test, app_name=PROM_NAME, unit_num=0)
+    assert len(rules) > 0, "No alert rules are present even though spring music is related"
+
+
+async def test_multiple_workloads_alert_rules(ops_test):
+    old_rules = await get_prometheus_rules(ops_test=ops_test, app_name=PROM_NAME, unit_num=0)
+    await ops_test.model.deploy(
+        "ch:spring-music", application_name="spring-music2", channel="edge"
+    )
```
Comment (sed-i) [inline tests/integration/test_charm.py:46]: Consider moving the `deploy` to the first function.

File: tests/unit/test_charm.py [hunk 9]
```diff
@@ -275,3 +275,63 @@ def test_unused_unit_sets_waiting_status_on_consumer_joined(self):
             self.harness.model.unit.status,
             WaitingStatus("inactive unit"),
         )
+
+    def test_alert_rules(self):
```
Comment (sed-i) [inline tests/unit/test_charm.py:279]: Consider having a dedicated test file for this, e.g. `test_relation_data.py`.
Even more so when `begin_with_initial_hooks()` is called in `setUp` because `set_leader` and `add_relation` [should be called beforehand](https://github.com/canonical/operator/blob/ff30d2f4bf0d8e2f24cd11d99e3de67d8e81e3b0/ops/testing.py#L213).

File: tests/unit/test_charm.py [hunk 10]
```diff
@@ -275,3 +275,63 @@ def test_unused_unit_sets_waiting_status_on_consumer_joined(self):
...
+        self.harness.update_relation_data(
+            workload_rel_id,
+            "cassandra-k8s",
+            {
+                "scrape_jobs": json.dumps(
+                    [{"metrics_path": "/metrics", "static_configs": [{"targets": ["*:9500"]}]}]
+                ),
+                "alert_rules": json.dumps(
+                    {
+                        "groups": [
```
Comment (sed-i) [inline tests/unit/test_charm.py:293]: Given this entire section is identical on both sides of the relation, why not reuse via a variable?

File: tests/unit/test_charm.py [hunk 11]
```diff
@@ -275,3 +275,63 @@ def test_unused_unit_sets_waiting_status_on_consumer_joined(self):
             self.harness.model.unit.status,
             WaitingStatus("inactive unit"),
         )
+
+    def test_alert_rules(self):
+        self.harness.set_leader(True)
+        prom_rel_id = self.harness.add_relation("metrics-endpoint", "prometheus-k8s")
+        workload_rel_id = self.harness.add_relation("configurable-scrape-jobs", "cassandra-k8s")
+        self.harness.add_relation_unit(workload_rel_id, "cassandra-k8s/0")
+        self.harness.update_relation_data(
```
Comment (sed-i) [inline tests/unit/test_charm.py:284]:
```suggestion

        # Set relation data on the "requires" side
        self.harness.update_relation_data(
```

File: tests/unit/test_charm.py [hunk 12]
```diff
@@ -275,3 +275,63 @@ def test_unused_unit_sets_waiting_status_on_consumer_joined(self):
...
+                                        },
+                                    }
+                                ],
+                            }
+                        ]
+                    }
+                ),
+            },
+        )
+        app_name = self.harness.model.app.name
```
Comment (sed-i) [inline tests/unit/test_charm.py:313]:
```suggestion
        
        # Verify relation data on the "provides" side matches the "requires" side
        app_name = self.harness.model.app.name
```

File: tests/unit/test_charm.py [hunk 13]
```diff
@@ -275,3 +275,63 @@ def test_unused_unit_sets_waiting_status_on_consumer_joined(self):
...
+        prom_rules = json.loads(
+            str(self.harness.get_relation_data(prom_rel_id, app_name).get("alert_rules"))
```
Comment (sed-i) [inline tests/unit/test_charm.py:315]: Add a similar one for `"scrape_jobs"`?

File: tests/unit/test_charm.py [hunk 14]
```diff
@@ -275,3 +275,63 @@ def test_unused_unit_sets_waiting_status_on_consumer_joined(self):
...
+                                    "juju_model_uuid": "test_uuid",
+                                    "juju_application": "test_app",
+                                    "juju_charm": "test_charm",
+                                },
+                            }
+                        ],
+                    }
+                ]
+            },
+        )
```
Comment (sed-i) [inline tests/unit/test_charm.py:337]: Add test case for empty input? I.e. the `else {}` in `alert_groups if alert_groups["groups"] else {}` etc.

File: tests/integration/test_charm.py (hunk 6 above)
Comment (dstathis) [reply tests/integration/test_charm.py:1]: Why not?

File: tests/integration/test_charm.py (hunk 7 above)
Comment (dstathis) [reply tests/integration/test_charm.py:29]: Why?

File: tests/unit/test_charm.py (hunk 13 above)
Comment (dstathis) [reply tests/unit/test_charm.py:315]: That is not what is being tested here.

File: tests/integration/test_charm.py (hunk 7 above)
Comment (sed-i) [reply tests/integration/test_charm.py:29]: - Increase clarity: all you need for this test can be deployed at the start, so it might be easier for the reader to understand all the actors at play.
- Follow up test funcs are (conveniently) named `test_relate` (and not `test_deploy_and_relate`).

This is just a suggestion so feel free to resolve.

File: tests/integration/test_charm.py (hunk 6 above)
Comment (dstathis) [reply tests/integration/test_charm.py:1]: Let me elaborate on why I think this is okay. If I were to change the name, it would likely be to `test_basic_functionality.py` because that's sort of what is going on. If someone wants to write a test which checks that the scrape jobs get set up properly (and someone should), they should not have to rewrite `test_build_and_deploy`.

File: tests/unit/test_charm.py (hunk 14 above)
Comment (dstathis) [reply tests/unit/test_charm.py:337]: Done.

File: tests/unit/test_charm.py (hunk 10 above)
Comment (dstathis) [reply tests/unit/test_charm.py]: Done.

File: tests/unit/test_charm.py (hunk 9 above)
Comment (dstathis) [reply tests/unit/test_charm.py:279]: The reason this is in test_charm is it is testing basic functionality. I don't really so a reason to rewrite all the boiler plate for something that makes sense to be in here.

Could you explain the issue with `set_leader` and and `add_relation` going after begin? Both of these things can happen after being deployed in production so why is it wrong to do it in test code?

Comment (rbarry82) [reply tests/unit/test_charm.py:279]: There is not one. This is not prescriptive, it's just an example in the docstring for how to hydrate data.

There is absolutely no problem in doing it afterwards to fire events as they come.

Comment (sed-i) [reply tests/unit/test_charm.py:279]: > set_leader and add_relation should be called beforehand

My bad, seems like that is not the case.

Comment (rbarry82) [reply tests/unit/test_charm.py:279]: I actually think it's a reasonable thought from reading the docstring and a good indication that it could be updated to at least reflect the fact that they are hydrating it, which could be useful in "racing" event firing.

**Review by rbarry82 — APPROVED** [review]

File: tests/integration/test_charm.py (hunk 6 above)
Comment (sed-i) [reply tests/integration/test_charm.py:1]: If we can sensibly fit all itests in one file then `test_charm` makes sense.

**Review by sed-i — DISMISSED** [review]
